From 9a1332b2697ecdc6a2c9f54737013c268e17b083 Mon Sep 17 00:00:00 2001
From: Bernt Budde <bernt.budde@gmail.com>
Date: Sat, 31 Oct 2026 05:17:41 +0200
Subject: [PATCH 82/85] TODO tests: <...> after a statement or map/grep block
 is a readline

After the } of a block that opens a statement or follows a word perl
expects a term, so <op/*>, <$fh> and <STDIN> there are a glob or a
readline (perl 5.40.3: my %h; my @l = map { $h{$_}++ } <DATA> reads
DATA).  PPI recognised only <WORD> after any }.  Two controls: after a
subscript's } and a do block's } the < compares.
---
 t/ppi_bugs_pcl.t | 30 +++++++++++++++++++++++++++++-
 1 file changed, 29 insertions(+), 1 deletion(-)

diff --git a/t/ppi_bugs_pcl.t b/t/ppi_bugs_pcl.t
index 79cf4fe1..dcac61a3 100644
--- a/t/ppi_bugs_pcl.t
+++ b/t/ppi_bugs_pcl.t
@@ -13,7 +13,7 @@
 use lib 't/lib';
 use PPI::Test::pragmas;
 use utf8;
-use Test::More tests => 266 + ($ENV{AUTHOR_TESTING} ? 1 : 0);
+use Test::More tests => 271 + ($ENV{AUTHOR_TESTING} ? 1 : 0);
 
 use PPI ();
 
@@ -1711,3 +1711,31 @@ TODO: {
     my $slurped = do { local $/; $tokens->() };
     is( $slurped, $plain, 'a standalone tokenizer: the tokens are the same when $/ is undef' );
 }
+
+# §14, the after-a-block half: after the } of a block that opens a
+# statement or follows a word perl expects a term, so <...> is a readline
+# or a glob whatever it holds.  perl: my %h; my @l = map { $h{$_}++ } <DATA>
+# reads DATA.  After a subscript's } or a do block's } (the controls) the
+# < compares.
+TODO: {
+    local $TODO = 'PCL §14: <FH> / <glob> after a list operator or a block is lexed as < … >';
+    for my $r (
+        [ 'my @l = map { $h{$_}++ } <op/*>;', 'QuoteLike::Readline=<op/*>',  'a glob after a map block' ],
+        [ 'my @m = grep { 1 } <$fh>;',        'QuoteLike::Readline=<$fh>',   'a readline after a grep block' ],
+        [ 'if ($x) { } <STDIN>;',             'QuoteLike::Readline=<STDIN>', 'a readline after a statement block' ],
+    ) {
+        my ( $code, $want, $desc ) = @$r;
+        my @t = map { short_class($_) } toks($code);
+        ok( grep( $_ eq $want, @t ), "$code: $desc" ) or diag "got: @t";
+    }
+}
+{
+    for my $r (
+        [ '$x = $h{a} < $b;',   'less-than after a subscript (control)' ],
+        [ '$w = do { 1 } < 2;', 'less-than after a do block (control)' ],
+    ) {
+        my ( $code, $desc ) = @$r;
+        my @t = map { short_class($_) } toks($code);
+        ok( grep( $_ eq 'Operator=<', @t ), "$code: $desc" ) or diag "got: @t";
+    }
+}
-- 
2.53.0

